### PR TITLE
feat(dyn-sampling): Automatic bias for Key Transactions

### PR DESCRIPTION
**Pull request by andriisoldatenko**
This PR implements BE for automatic bias for Key Transactions

When user select boosting key transactions we invalidate project config cache if dynamic-sampling
feature flag is enabled.

Also we limit key_transactions  up to `BOOSTED_KEY_TRANSACTION_LIMIT` and currently it set to `10`.
So we don't send to relay too many transaction patterns per project.

Also currently there is no proper way to order key transaction by: `created` or `updated` so it's just random unique `10` for give project.

```
sentry=# select * from sentry_performanceteamkeytransaction;
 id |                   transaction                    | organization_id | project_team_id
----+--------------------------------------------------+-----------------+-----------------
  2 | /api/0/organizations/{organization_slug}/events/ |               1 |               1
  3 | sentry.tasks.send_ping                           |               1 |               1
  4 | /heartbeat3                                      |               1 |              11
  5 | /heartbeat2                                      |               1 |              11
```

Also important note: we select 10 unique transaction for all teams for given project_id. So if 2 teams select 2 same transactions we will boost only single unique one.

It depends on ~https://github.com/getsentry/relay/pull/1581~  :merged 

FE PR -> https://github.com/getsentry/sentry/pull/41392

### CODE REVIEW

**Comment by andriisoldatenko** [comment]
cc @imatwawana

**Review by jjbayer — APPROVED** [review]

File: src/sentry/discover/models.py [hunk 1]
```diff
@@ -76,6 +85,53 @@ def set_projects(self, project_ids):
             )
 
 
+class TeamKeyTransactionModelManager(BaseManager):
+    @staticmethod
+    def __schedule_invalidate_project_config_transaction_commit(instance, trigger):
+        project = (
+            instance.project_team.project if hasattr(instance.project_team, "project") else None
+        )
```
[the remark this answers is not in the document]
Comment (andriisoldatenko) [reply src/sentry/discover/models.py:93]: i saw it in sentry errors. /shrug

File: src/sentry/dynamic_sampling/rules_generator.py [hunk 2]
```diff
@@ -178,6 +203,18 @@ def generate_rules(project: Project) -> List[Union[BaseRule, ReleaseRule]]:
...
+            # Key Transaction boost
+            if RuleType.BOOST_KEY_TRANSACTIONS_RULE.value in enabled_biases:
+                key_transactions = list(
+                    TeamKeyTransaction.objects.filter(
+                        organization_id=project.organization.id,
+                        project_team__project_id=project.id,
+                    )
+                    .values_list("transaction", flat=True)
+                    .distinct()[:BOOSTED_KEY_TRANSACTION_LIMIT]
+                )
```
Comment (iambriccardo) [inline src/sentry/dynamic_sampling/rules_generator.py:215]: I don't like having logic put inside here. In `generate_boost_release_rules` we also have logic to fetch data and I would prefer to keep implementations similar. In addition, this adds complexity to this function that is out of scope.

In the future I would suggest to rework the implementation as discussed privately.

Comment (andriisoldatenko) [reply src/sentry/dynamic_sampling/rules_generator.py:215]: yeah i agree, will extract to utils this block.

File: src/sentry/discover/models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (andriisoldatenko) [reply src/sentry/discover/models.py]: fixed. cc @jjbayer

File: src/sentry/dynamic_sampling/rules_generator.py (hunk 2 above)
Comment (andriisoldatenko) [reply src/sentry/dynamic_sampling/rules_generator.py]: fixed @iambriccardo

**Review by iambriccardo — APPROVED** [review]

**Comment by andriisoldatenko** [comment]
related docs PR https://github.com/getsentry/sentry-docs/pull/5833

**Comment by andriisoldatenko** [comment]
practically it also works:
```
(.venv) ➜  demo git:(main) ✗ python get_projectconfig.py --public-key=<> | jq
{
  "rules": [
    {
      "sampleRate": 1,
      "type": "transaction",
      "condition": {
        "op": "or",
        "inner": [
          {
            "op": "glob",
            "name": "event.transaction",
            "value": [
              "/api/0/organizations/{organization_slug}/stats/",
              "/api/0/projects/{organization_slug}/{project_slug}/releases/"
            ],
            "options": {
              "ignoreCase": true
            }
          }
        ]
      },
      "active": true,
      "id": 1003
    },
    {
      "sampleRate": 0.5,
      "type": "trace",
      "active": true,
      "condition": {
        "op": "and",
        "inner": []
      },
      "id": 1000
    }
  ]
}
```

image

**Review by jjbayer — CHANGES_REQUESTED** [review]
Relay interprets `*`, `[` and `{` as special directives in glob patterns, so we would have to escape them if they occur in transaction names. We should use the [`eq` "op"](https://github.com/getsentry/relay/blob/df231510f7209c427e8e4a7038dd31ce65e47f1f/tests/integration/test_outcome.py#L621-L625) instead.

See failing test case here: https://github.com/getsentry/relay/pull/1620

Credit goes to @andriisoldatenko for finding this.

**Comment by iambriccardo** [comment]
> Relay interprets `*`, `[` and `{` as special directives in glob patterns, so we would have to escape them if they occur in transaction names. We should use the [`eq` "op"](https://github.com/getsentry/relay/blob/df231510f7209c427e8e4a7038dd31ce65e47f1f/tests/integration/test_outcome.py#L621-L625) instead.
> 
> See failing test case here: [getsentry/relay#1620](https://github.com/getsentry/relay/pull/1620)
> 
> Credit goes to @andriisoldatenko for finding this.

I assume we will also have to handle this anytime we send you a string that is matched with the `glob` operator.

**Comment by iambriccardo** [comment]
@jjbayer what do `{` and `[` do as special directives?

**Comment by priscilawebdev** [comment]
tested it locally and everything seems to work fine 


https://user-images.githubusercontent.com/29228205/203380788-ebd59a9a-4583-410b-b296-707865bb1eec.mov

**Comment by andriisoldatenko** [comment]
Thanks!

File: src/sentry/dynamic_sampling/utils.py [hunk 3]
```diff
@@ -11,13 +14,14 @@ class Bias(TypedDict):
...
-# These represent the biases that are applied to user by default as part of the adaptive dynamic sampling experience.
-# These can be overridden by the project details endpoint
+# These represent the biases that are applied to user by default as part of the adaptive dynamic sampling
+# experience. These can be overridden by the project details endpoint
 class RuleType(Enum):
     UNIFORM_RULE = "uniformRule"
     BOOST_ENVIRONMENTS_RULE = "boostEnvironments"
     BOOST_LATEST_RELEASES_RULE = "boostLatestRelease"
     IGNORE_HEALTHCHECKS_RULE = "ignoreHealthChecks"
+    BOOST_KEY_TRANSACTIONS_RULE = "boostKeyTransaction"
```
[the remark this answers is not in the document]
Comment (andriisoldatenko) [reply src/sentry/dynamic_sampling/utils.py:24]: @priscilawebdev fixed, please update UI PR too.

**Comment by andriisoldatenko** [comment]
> Relay interprets `*`, `[` and `{` as special directives in glob patterns, so we would have to escape them if they occur in transaction names. We should use the [`eq` "op"](https://github.com/getsentry/relay/blob/df231510f7209c427e8e4a7038dd31ce65e47f1f/tests/integration/test_outcome.py#L621-L625) instead.
> 
> See failing test case here: [getsentry/relay#1620](https://github.com/getsentry/relay/pull/1620)
> 
> Credit goes to @andriisoldatenko for finding this.

fixed.

File: src/sentry/discover/models.py [hunk 4]
```diff
@@ -76,6 +85,51 @@ def set_projects(self, project_ids):
             )
 
 
+class TeamKeyTransactionModelManager(BaseManager):
+    @staticmethod
+    def __schedule_invalidate_project_config_transaction_commit(instance, trigger):
+        project = getattr(instance.project_team, "project")
```
Comment (jjbayer) [inline src/sentry/discover/models.py:91]: I'm confused about this: `getattr` with two arguments will raise an `AttributeError` if `project` does not exist, so we might as well write `instance.project_team.project`. Unless we do not want it to raise an exception, in which case we should use `getattr(a, b, None)`.

https://docs.python.org/3/library/functions.html#getattr

**Review by jjbayer — APPROVED** [review]
See question about `getattr`, apart from that this looks good to me!

File: src/sentry/discover/models.py (hunk 4 above)
Comment (andriisoldatenko) [reply src/sentry/discover/models.py:91]: oh you are right!
```
In [1]: class Foo:
   ...:     pass
   ...:

In [2]: f = Foo()

In [3]: getattr(f, "porject")
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [3], in <cell line: 1>()
----> 1 getattr(f, "porject")

AttributeError: 'Foo' object has no attribute 'porject'

In [4]: getattr(f, "porject", None)

In [5]:
```

Comment (andriisoldatenko) [reply src/sentry/discover/models.py:91]: @jjbayer fixed!

**Review by priscilawebdev — APPROVED** [review]

**Pull request merged**